src/bucket/BucketList.cpp

```diff
@@ -371,9 +377,227 @@ BucketList::getHash() const
     return hsh.finish();
 }
 
-// levelShouldSpill is the set of boundaries at which each level should spill,
-// it's not-entirely obvious which numbers these are by inspection, so we list
-// the first 3 values it's true on each level here for reference:
+void
+BucketList::loopAllBuckets(std::function<bool(std::shared_ptr<Bucket>)> f) const
+{
+    for (size_t i = 0; i < mLevels.size(); ++i)
+    {
+        auto const& lev = mLevels[i];
+        std::array<std::shared_ptr<Bucket>, 2> buckets = {lev.getCurr(),
+                                                          lev.getSnap()};
+        for (auto b : buckets)
+        {
+            // Once we see one empty bucket, the rest will also be empty, return
+            // early
+            if (b->getFilename().empty())
```

Comment:
Correct

Reply:
Yeah, older versions buckets are weird because they don't contain METAENTRY. So we should be conservative and always iterate over the whole bucketslist. 

Also - I think it's error-prone to check filenames like this. The caller isn't supposed to know an internal to Bucket detail that empty buckets have empty filenames. What we should do is have a public method in the Bucket class, which checks whether the bucket is empty. While we're there, we should check that both filename and hash are empty (otherwise throw, as it's an invalid bucket). We should clean up the code to use that function everywhere: from what I see, we also check for empty buckets in BucketList::restartMerges and in HistoryArchiveState::containsValidBuckets.

Reply:
thanks for cleaning this up!